### PR TITLE
Standard naming for Python C extensions

### PR DESCRIPTION
**Pull request by bjones1**
This PR enables PyInstaller to follow the standard Python naming convention for Python C extensions, by placing them at `module/submodule/extension_name.ext`, rather than `module.submodule.extension_name.ext`. Following the standard Python naming convention lowers the maintenance burden, by allowing PyInstaller to more closely mimic Python's behavior. In particular, the ZMQ hook becomes much simpler.

Note that these changes only apply to Python 3; issues in Python 2's import system prevent it from being used.

@htgoebel, I'd like you to approve before proceeding.

### CODE REVIEW

File: PyInstaller/building/utils.py [hunk 1]
```diff
@@ -95,11 +95,24 @@ def add_suffix_to_extensions(toc):
...
             # In some rare cases extension might already contain a suffix.
             # Skip it in this case.
             if os.path.splitext(inm)[1] not in EXTENSION_SUFFIXES:
-                # Use this file's existing extension.
-                inm = inm + os.path.splitext(fnm)[1]
+                # Determine the base name of the file.
+                if is_py3:
+                    base_name = os.path.basename(inm)
+                else:
+                    base_name = inm.rpartition('.')[2]
```
Comment (htgoebel) [inline PyInstaller/building/utils.py]: Use `.rsplit()`?

File: PyInstaller/building/utils.py [hunk 2]
```diff
@@ -95,11 +95,24 @@ def add_suffix_to_extensions(toc):
     new_toc = TOC()
     for inm, fnm, typ in toc:
         if typ == 'EXTENSION':
+            if is_py3:
+                # Change the dotted name into a relative path. This places C
+                # extensions in this Python-standard location. This only works
```
Comment (htgoebel) [inline PyInstaller/building/utils.py]: s/in this/in the/ ?

File: PyInstaller/building/utils.py [hunk 3]
```diff
@@ -95,11 +95,24 @@ def add_suffix_to_extensions(toc):
...
             # Skip it in this case.
             if os.path.splitext(inm)[1] not in EXTENSION_SUFFIXES:
-                # Use this file's existing extension.
-                inm = inm + os.path.splitext(fnm)[1]
+                # Determine the base name of the file.
+                if is_py3:
+                    base_name = os.path.basename(inm)
+                else:
+                    base_name = inm.rpartition('.')[2]
+                assert '.' not in base_name
```
Comment (htgoebel) [inline PyInstaller/building/utils.py]: Leftover? Reading the code, this can not happen-

File: PyInstaller/building/utils.py [hunk 4]
```diff
@@ -95,11 +95,24 @@ def add_suffix_to_extensions(toc):
...
+                # Use this file's existing extension. For extensions such as ``libzmq.cp36-win_amd64.pyd``, we can't use ``os.path.splitext``, which would give only the ```.pyd`` part of the extension.
```
Comment (htgoebel) [inline PyInstaller/building/utils.py]: Interesting point!

Please wrap the long line.

File: PyInstaller/loader/pyimod02_archive.py [hunk 5]
```diff
@@ -259,7 +259,6 @@ def __init__(self):
             self.key = key.zfill(CRYPT_BLOCK_SIZE)
         assert len(self.key) == CRYPT_BLOCK_SIZE
 
-        # Import the right AES module.
```
Comment (htgoebel) [inline PyInstaller/loader/pyimod02_archive.py]: IMHO this comment is useful.

File: PyInstaller/loader/pyimod02_archive.py [hunk 6]
```diff
@@ -268,23 +267,33 @@ def _import_aesmod(self):
...
+                    raise ImportError(modname)
+            mod = mod.load_module(modname)
+        else:
+            kwargs = dict(fromlist=['Crypto', 'Cipher'])
+            try:
+                mod = __import__(modname, **kwargs)
+            except ImportError:
+                modname = 'Crypto.Cipher.AES'
+                mod = __import__(modname, **kwargs)
+
```
Comment (htgoebel) [inline PyInstaller/loader/pyimod02_archive.py]: How is this change related to the commit? Look like this is fixing some other issue. If so, please put it into a commit of its own

File: tests/functional/test_import.py [hunk 7]
```diff
@@ -172,13 +172,6 @@ def test_import_non_existing_raises_import_error(pyi_builder):
             raise RuntimeError("ImportError not raised")
         """)
 
-# :todo: Use some package which is already installed for some other
-# reason instead of `simplejson` which is only used here.
-@importorskip('simplejson')
```
Comment (htgoebel) [inline tests/functional/test_import.py]: What do you think about keeping this for Python2? Of course then the Py3-specific code can be removed from the test-script.

File: PyInstaller/hooks/hook-zmq.py [hunk 8]
```diff
@@ -17,18 +17,19 @@
...
-                 '.' if is_py2 and is_win else 'zmq')]
-except ImportError:
-    # Not all platforms provide their own copy of libzmq.
-    pass
-else:
-    # If PyZMQ pvorides its own copy of libzmq and libsodium, these libs look like
-    # C extensions. Excluding these modules ensures that those dlls are not bundled
-    # twice. Once as ./zmq.libzmq.pyd and once as ./zmq/libzmq.py.
-    excludedimports = ['zmq.libzmq']
+if is_py2:
```
Comment (htgoebel) [inline PyInstaller/hooks/hook-zmq.py]: Please add a comment, why this is only required for Py2.

File: PyInstaller/hooks/hook-zmq.py [hunk 9]
```diff
@@ -17,18 +17,19 @@
...
+    # If PyZMQ provides its own copy of libzmq and libsodium, add it to the
+    # extension-modules TOC so zmq/__init__.py can load it at runtime.
+    # PyZMQ is able to load 'libzmq' and 'libsodium' even from sys._MEIPASS.
+    # So they could be with other .dlls.
```
Comment (htgoebel) [inline PyInstaller/hooks/hook-zmq.py]: should this be "like they could do with other .dlls"?

File: PyInstaller/loader/pyimod03_importers.py [hunk 10]
```diff
@@ -234,7 +234,7 @@ def __init__(self):
 
     def __call__(self, path):
         """
-        PEP-302 sys.path_hook processor.
+        PEP-302 sys.path_hook processor. Only needed for Python 2.
```
Comment (htgoebel) [inline PyInstaller/loader/pyimod03_importers.py]: Please explain briefly why only used for Python 2.

File: PyInstaller/building/utils.py (hunk 1 above)
Comment (bjones1) [reply PyInstaller/building/utils.py]: OK, done.

File: PyInstaller/building/utils.py (hunk 2 above)
Comment (bjones1) [reply PyInstaller/building/utils.py]: Oops, thanks! Fixed.

File: PyInstaller/building/utils.py (hunk 3 above)
Comment (bjones1) [reply PyInstaller/building/utils.py]: True, just an assertion to make sure my code didn't behave in any unexpected ways. That is, `os.path.basename` should never report a filename containing a period. If so, something is very wrong.

File: PyInstaller/building/utils.py (hunk 4 above)
Comment (bjones1) [reply PyInstaller/building/utils.py]: Found only by experimentation -- this surprised me too. Wrapped.

File: PyInstaller/loader/pyimod02_archive.py (hunk 5 above)
Comment (bjones1) [reply PyInstaller/loader/pyimod02_archive.py]: Good point, restored.

File: PyInstaller/loader/pyimod02_archive.py (hunk 6 above)
Comment (bjones1) [reply PyInstaller/loader/pyimod02_archive.py]: The Python 2 code uses the `CExtensionImporter` to import `Crypto.Cipher.AES`; Python 3 can't, since the extension can now be imported using the standard Python import mechanism. Hence this change.

I put this in the same commit, since these two changes (removing the `CExtensionImporter` and removing code which uses `CExtensionImporter`) are related.

File: tests/functional/test_import.py (hunk 7 above)
Comment (bjones1) [reply tests/functional/test_import.py]: Good point. I've kept this test, but now it only runs on Python 2.

File: PyInstaller/hooks/hook-zmq.py (hunk 9 above)
Comment (bjones1) [reply PyInstaller/hooks/hook-zmq.py]: Possibly? This is a copy-and-paste from the existing hook. This comment makes no sense to me, and I don't even understand what it's trying to say. I'll go ahead and make your edit.

File: PyInstaller/hooks/hook-zmq.py (hunk 8 above)
Comment (bjones1) [reply PyInstaller/hooks/hook-zmq.py]: Done.

File: PyInstaller/loader/pyimod03_importers.py (hunk 10 above)
Comment (bjones1) [reply PyInstaller/loader/pyimod03_importers.py]: Done.

**Comment by bjones1** [comment]
I'm not sure why I closed this. It's almost done. I can't seem to reopen it, either.

**Comment by bjones1** [comment]
I see -- I changed the branch name, perhaps? In any case, I'll merge this if the tests pass.

**Comment by htgoebel** [comment]
Please rebase-merge, I try to avoid unnecessary merges. Thanks.
